cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,309 @@
+    """
+
+    def __init__(
+        self,
+        *gates: Union[Type[raw_types.Gate], raw_types.Gate, GateFamily],
+        name: Optional[str] = None,
+    ) -> None:
+        """Init Gateset.
+
+        Accepts a list of gates, each of which should be either
+            a) `cirq.Gate` subclass
+            b) `cirq.Gate` instance
+            c) `cirq.GateFamily` instance
+
+        `cirq.Gate` subclasses and instances are converted to the default
+        `cirq.GateFamily(gate=g)` instance and thus a default name and
+        description is populated.
+
+        Args:
+            *gates: A list of `cirq.Gate` subclasses / `cirq.Gate` instances /
+            `cirq.GateFamily` instances to initialize the Gateset.
+            name: (Optional) Name for the Gateset. Useful for description.
+        """
+        self._name = name
+        self._gates = frozenset(
+            g if isinstance(g, GateFamily) else GateFamily(gate=g) for g in gates
+        )
+        self._instance_gate_families: Dict[raw_types.Gate, GateFamily] = {}
+        self._type_gate_families: Dict[Type[raw_types.Gate], GateFamily] = {}
+        self._custom_gate_families: List[GateFamily] = []
+        for g in self._gates:
+            if type(g) == GateFamily:
+                if isinstance(g.gate, raw_types.Gate):
+                    self._instance_gate_families[g.gate] = g
+                else:
+                    self._type_gate_families[g.gate] = g
+            else:
+                self._custom_gate_families.append(g)
+
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
+    @property
+    def gates(self) -> FrozenSet[GateFamily]:
+        return self._gates
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        """Check for containment of `item` in any of the member `GateFamily`.
+
+        This method just forwards containment checks to the underlying
+        `GateFamily` objects. It returns `True` if any of the member gate
+        families can accept `item`.
+
+        To validate whether a `cirq.CircuitOperation`/`cirq.OP_TREE`/
+        `cirq.Circuit` is made up entirely of gates from this Gateset, use the
+        `validate` method given below.
+
+        The complexity of the method is:
+            a) O(1) for checking containment in the default `cirq.GateFamily` instances.
+            b) O(n) for checking containment in custom GateFamily instances.
+
+        Args:
+            item: The `cirq.Gate` or `cirq.Operation` instance to check
+                containment for.
+        """
+        g = item if isinstance(item, raw_types.Gate) else item.gate
+        if g in self._instance_gate_families:
+            assert item in self._instance_gate_families[g]
+            return True
+        for gate_mro_type in type(g).mro():
+            if gate_mro_type in self._type_gate_families:
+                assert item in self._type_gate_families[gate_mro_type]
```

Comment:
Nit: remove assert statements.

Reply:
I think the assert here is important to verify that the internal code is working as expected. It's a tricky implementation and the assert would be useful to safeguard in any future changes we might make.

Reply:
Fair enough: Can you add a description to them if they get triggered that reads something like: "Match found in GateSet with missing underlying GateFamily entry" for each of the asserts then ?

Reply:
done.